src/execnet/gateway_base.py

```diff
@@ -1106,7 +1129,18 @@ def join(self, timeout=None):
 class WorkerGateway(BaseGateway):
     def _local_schedulexec(self, channel, sourcetask):
         sourcetask = loads_internal(sourcetask)
-        self._execpool.spawn(self.executetask, (channel, sourcetask))
+        if self.execmodel.backend == "main_thread_only":
+            # TODO: Maybe use something like queue.Queue to queue an asynchronous
+            # spawn here in order to avoid using another thread.
+            import threading
+
+            t = threading.Thread(
+                target=self._execpool.spawn,
+                args=(self.executetask, (channel, sourcetask)),
+            )
+            t.start()
+        else:
+            self._execpool.spawn(self.executetask, (channel, sourcetask))
```

Comment:
Apparently TestBasicGateway::test_gateway_status_busy doesn't deadlock (as reported [previously](https://github.com/pytest-dev/execnet/pull/243#issuecomment-1945525625)) when behavior is preserved for the thread execmodel.

Reply:
It works because hasprimary is False for the receivepool, so _primary_thread_task_ready is always None for this pool and _thread_receiver runs in a non-main thread despite the main_thread_only execmodel.

This is great because it allows us to always call `self._execpool.spawn` directly here, consistently providing backpressure to _thread_receiver.

@RonnyPfannschmidt without the threading.Thread hack here, it operates essentially as before, so I guess we don't need a special timeout test for it anymore? I imagine the backpressure should propagate back to the original caller when _thread_receiver is blocking on spawn, and _thread_receiver will catch any exception and propagate at back to the orginal caller as well.

Reply:
I added a test to make sure we can submit several `remote_exec` calls in quick succession and have them all execute in the main thread:
```python
    # Create multiple channels at once and assert that all tasks
    # execute in the main thread.
    channels = []
    for i in range(10):
        channels.append(
            gw.remote_exec(
                """
                import time, threading
                time.sleep(0.02)
                channel.send(threading.current_thread() is threading.main_thread())
        """
            )
        )

    for ch in channels:
        res = ch.receive()
        ch.close()
        if res is not True:
            pytest.fail("remote raised\n%s" % res)
```

Reply:
Execnet is completely missing a explicit Handling of back pressure

Not adding a error is likely to trigger a problem unexpectedly

Reply:
The key here is that the main thread only models practically dreadlocks on the contention of the main thread, shooting up random threads to pretend sidestepping that problem is a tech debt I'm not keen on taking in

Error early instead

Reply:
I've already removed the "random threads" though, since I was unable to trigger a deadlock after having split out the main_thread_only execmodel that inherits from ThreadExecModel. We could trigger an error if we detect that there is more than one self._execpool.spawn call in flight at once, but that can't happen now because the single _thread_receiver thread would be busy executing the first call to self._execpool.spawn, triggering back pressure since any pending requests would accumulate in self._io. So, the natural place to trigger an error would be inside self._io if we detect more than one concurrent request accumulating there and not being serviced due to back pressure (`test_assert_main_thread_only` will trigger this error if IO buffering prevents remote_exec from blocking). In test_assert_main_thread_only we can avoid triggering this error by waiting for previous channel responses before sending new remote_exec requests.

In order to detect accumulating requests in self._io, we'll need one additional thread, or else we'll have to convert _thread_receiver to an event loop that is capable of servicing self._io while it waits for previous asynchronous calls to self._execpool.spawn to acquire the main thread.

Alternatively, we could alter the remote_exec protocol to require a response that creates back pressure, and then remote_exec would naturally block for this response when the _thread_receiver thread is busy acquiring the main thread in self._execpool.spawn. So this makes three choices:

* Use an additional thread to detect accumulating requests in self._io and trigger an error.
* Use an event loop running in the _thread_receiver thread to detect accumulating requests in self._io, and also implement asynchronous self._execpool.spawn calls that this event loop can monitor concurrently with self._io.
* Change the remote_exec protocol to require a response that creates back pressure when the _thread_receiver thread is busy acquiring the main thread in self._execpool.spawn.

Reply:
@RonnyPfannschmidt the "use an additional thread to detect accumulating requests in self._io and trigger an error" option seems the least invasive, and it seems best if this new behavior is conditional on the main_thread_only execmodel, so that behavior for existing execmodels is preserved (accumulating requests are less likely to be an issue since they don't need to acquire the main thread).

Reply:
I suppose we could possibly avoid using a separate thread to monitor for accumulating requests in self._io if we make _thread_receiver use non-blocking IO to read all buffered requests and then raise an error if there is more than one buffered request.

Reply:
The non-blockin IO approach could still lead to a deadlock (before we have detected accumulating requests) if a previous request is executing in the main thread and waiting for some condition that a later request is intended to satisfy, but this behavior could be acceptable since the user will notice the deadlock and have to solve it somehow (and this isn't a breaking change since main_thread_only is a new and non-default execmodel).

Reply:
Actually, we should be able to use non-blocking IO to detect the accumulation before it deadlocks if we refrain from trying to call self._execpool.spawn while a previous request is still running in the main thread.

Reply:
We could use the stdlib [selectors](https://docs.python.org/3/library/selectors.html) API (available since python 3.4) for a non-blocking IO loop, and the loop won't starve much if it avoids calling self._execpool.spawn while a previous request is still running in the main thread. We can add some kind of callback to wake up the IO loop when the main thread isn't busy. The callback could just write an arbitrary single byte into a pipe that the IO loop is monitoring.

Reply:
We can use a separate _thread_receiver like method to implement the non-blocking IO loop just for the main_thread_only execmodel, and keep using the existing (blocking) _thread_receiver implementation for the existing execmodels which have no need to acquire the main thread.

Reply:
Can we just drop the worker pool/spawn for the main thread only case and simply return a error to the channel?

Reply:
I suppose at this point we could possibly remove support for waiting for the previous task in  _try_send_to_primary_thread, but it could still be useful if there's ever a context where one would want to submit multiple spawn calls and have them all run serially in the main thread, with spawn blocking to provide back pressure. I audited the internal spawn calls and it seems like I've found one such context in the `safe_terminate` function where it calls spawn in a loop:
```python
def safe_terminate(execmodel, timeout, list_of_paired_functions):
    workerpool = WorkerPool(execmodel)

    def termkill(termfunc, killfunc):
        termreply = workerpool.spawn(termfunc)
        try:
            termreply.get(timeout=timeout)
        except OSError:
            killfunc()

    replylist = []
    for termfunc, killfunc in list_of_paired_functions:
        reply = workerpool.spawn(termkill, termfunc, killfunc)
        replylist.append(reply)
    for reply in replylist:
        reply.get()
    workerpool.waitall(timeout=timeout)
```